### ISSUE TITLE
allowSingleLine in allman style for brace-style

### ISSUE THREAD

**Issue by PhiLhoSoft** [eslint/eslint#5197]
I use the Allman style of braces, so I have this setting in my .eslintrc.yaml file:
  
```
brace-style:
    - 1
    - allman
    - allowSingleLine: true
```
If I have this setting, brace style checking is entirely deactivated!
If I replace `allman` with `stroustrup`, it works, reporting lines with Allman style and not those in Stroustrup style, even on one line.
If I remove the last line, it works, but it complains on single line braces.

Sometime, eg. when making small objects like `{ id: 'foo' }`, I want to use the single line style.
Is it forbidden for the Allman style? Why?

**Comment by eslintbot** [comment]
@PhiLhoSoft Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
@PhiLhoSoft Thanks for the report. Could you provide information requested by @eslintbot please? Especially a few short code examples and output from ESLint.

**Comment by PhiLhoSoft** [comment]
Yes, I was thinking about this while coming here for updates...
A test case should be easy to set up.
I will come back with one soon...
Thanks for the quick response.

**Comment by PhiLhoSoft** [comment]
OK, I made a fairly complete testbench... :smile: 
[TestESLint.zip](https://github.com/eslint/eslint/files/124715/TestESLint.zip)

I wrote the config files both in Json and in Yaml format, because I was unsure of the Yaml format, so I wanted to check if results with both formats are the same. That's the case.

These tests shown the real issue with the Allman style check: if allowSingleLine is true, it almost accepts anything!
It only complains on a `} else {` construct...

I think that if aLS is true in Allman style, it should do an additional check: if the opening brace is on the same line than the controlling statement, then the closing brace must be on the same line too. Otherwise, complain!

Also, I was expecting these rules to be applies to objects and array literals, but that's not the case.
I will check if there are separate rules for them, otherwise I will open a new issue...
At least; I suggest to add to the documentation of `brace-style` that it is not used on object literals.

**Comment by nzakas** [comment]
@PhiLhoSoft can you paste a simple repro-case directly into the issue? A lot of us respond to issues from phone or tablet, so downloading a file means we can't respond until at a computer. If you can put the information directly into a comment, then we can move faster.

**Comment by PhiLhoSoft** [comment]
The essence of my tests is: the setting works in all cases, except on Allman with allowSingleLine, because it doesn't check that braces are aligned in a single line case.

Rule:

```
rules:
  brace-style:
    - 1
    - allman
    - allowSingleLine: true
```
(or `"brace-style": [ 1, "allman", { "allowSingleLine": true } ]`)

Test file:
```
function foo(param)
{
	var result;
	if (param)
	{
		result = { x: 1, y: 2, for: bar() };
	}
	else
	{
		result =
		{
			horizontal: true,
			vertical: 'maybe',
			for: bar()
		};
	}
	return {
		type: 'foo',
		payload: result
	};
}

// Single line, on line 25

function bar() { return 'bar'; }

// Bad styles, OK / expected after line 25...

function bad(b) { // Should complain
	var result;
	if (b){ // Should complain
		result = {};
	}
	else { // Should complain
		result = [];
	}
	return result;
}
```

**Comment by PhiLhoSoft** [comment]
Proposed fix: in checkBlock(), replace
```
                if (style !== "allman" && previousToken.loc.start.line !== curlyToken.loc.start.line) {
                    context.report(node, OPEN_MESSAGE);
                } else if (style === "allman" && previousToken.loc.start.line === curlyToken.loc.start.line && !params.allowSingleLine) {
                    context.report(node, OPEN_MESSAGE_ALLMAN);
                }
```
with
```
                if (style !== "allman" && previousToken.loc.start.line !== curlyToken.loc.start.line) {
                    context.report(node, OPEN_MESSAGE);
                } else if (style === "allman" && previousToken.loc.start.line === curlyToken.loc.start.line) {
                    if (params.allowSingleLine) {
                        if (curlyTokenEnd.loc.start.line !== curlyToken.loc.start.line) {
                            context.report(node, OPEN_MESSAGE_ALLMAN);
                        }
                    } else {
                        context.report(node, OPEN_MESSAGE_ALLMAN);
                    }
                }
```
I will try and make a proper PR...

**Comment by aparajita** [comment]
I can confirm this fix works. Here is a complete set of tests. All but the first should generate an error:

```js
/* eslint brace-style: [2, "allman", { "allowSingleLine": true }] */

function foo() { console.log("foo"); }

function bar() {
    doSomething();
}

if (foo) {
    console.log("bar");
}

if (foo)
{
    doSomething();
}
else {
    doSomethingElse();
}
```

**Comment by nzakas** [comment]
@PhiLhoSoft @aparajita can one of you open a pull request?

**Comment by PhiLhoSoft** [comment]
Yes, I try to work on it, on my scarce free time... :smile: But if @aparajita is faster than me, I won't mind at all! :wink: 
Note: the test cases should be a bit more comprehensive. That's the necessary, time-consuming point, I think.

**Comment by BYK** [comment]
@PhiLhoSoft, @aparajita - I took this on since there wasn't much activity. Sorry if I stepped on your toes! :)

**Comment by aparajita** [comment]
@BYK Be my guest!

**Comment by PhiLhoSoft** [comment]
No problem, I appreciate the push! Hadn't time to do it anyway.
